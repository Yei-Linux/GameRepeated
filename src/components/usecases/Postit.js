import { Fragment } from 'react'
import classNames from 'classnames'
import Container from '../ui/layouts/Container'
import FormItem from '../ui/atoms/FormItem'
import { useExerciseStore } from '../../hooks/useExerciseStore'

const Postit = ({
  text,
  theme = 'primary',
  readonly = true,
  name,
  htmlFor,
  ariaLabel,
}) => {
  const { solution, setSolution } = useExerciseStore()

  return (
    <Fragment>
      {readonly && text !== undefined && (
        <Container
          testId="postitReadonly"
          className={classNames(
            'postit',
            'postit--dimensions',
            `postit--${theme}`,
            'postit--readonly'
          )}
        >
          {text}
        </Container>
      )}
      {!readonly && (
        <FormItem
          testId="postitNotReadonly"
          htmlFor={htmlFor}
          ariaLabel={ariaLabel}
          value={solution[name]}
          name={name}
          onChange={({ value }) => {
            setSolution(value, name)
          }}
          className={classNames('postit--notreadonly')}
          classNameInput={classNames(
            'postit--dimensions',
            'postit__inputnumber--dimensions'
          )}
        />
      )}
    </Fragment>
  )
}

export default Postit
