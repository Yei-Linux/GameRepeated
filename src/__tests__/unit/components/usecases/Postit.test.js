import { render, screen } from '@testing-library/react'
import Postit from '../../../../components/usecases/Postit'

import ExerciseContext from '../../../../store/context/Exercise/context'
import userEvent from '@testing-library/user-event'

import ExerciseStore from '../../../../store/context/Exercise/store'

describe('The Postit component', () => {
  const defaultProps = {
    htmlFor: 'postItElementLabel',
    ariaLabel: 'postItElement',
    text: 'PostIt Component',
    theme: 'primary',
    readonly: true,
    name: '0',
  }
  describe('Postit test with mocked context', () => {
    const events = {
      setExercise: jest.fn(),
      setExerciseType: jest.fn(),
      setSolution: jest.fn(),
      setWaitMs: jest.fn(),
      setIsVisibleQuestion: jest.fn(),
    }
    const values = {
      isVisibleQuestion: true,
      waitMs: 5000,
      exercise: [],
      type: 'numbers',
      solution: Array.apply(null, new Array(5)).reduce(
        (acc, _, currentIndex) => ({
          ...acc,
          [`${currentIndex}`]: 'abc',
        }),
        {}
      ),
    }

    const setupComponent = (props = defaultProps) =>
      render(
        <ExerciseContext.Provider
          value={{
            ...events,
            ...values,
          }}
        >
          <Postit {...props} />
        </ExerciseContext.Provider>
      )

    describe('Postit readonly component', () => {
      it('Should render Postit component of readonly type', () => {
        const { asFragment } = setupComponent()

        expect(asFragment()).toMatchSnapshot()
      })
      it('Should make readonly the postit component', () => {
        const { text } = defaultProps
        setupComponent()

        const postit = screen.getByText(text)

        expect(postit).toHaveClass('postit--readonly')
      })
    })

    describe('Postit not readonly component', () => {
      beforeEach(() => {
        setupComponent({
          ...defaultProps,
          text: undefined,
          readonly: false,
        })
      })

      it('Should render Postit component of not readonly type', () => {
        const { asFragment } = setupComponent({
          ...defaultProps,
          text: undefined,
          readonly: false,
        })

        expect(asFragment()).toMatchSnapshot()
      })

      it('Should render label and input from postit', () => {
        const { htmlFor, ariaLabel } = defaultProps

        const labelPostit = screen.queryByLabelText(htmlFor)
        const inputPostit = screen.getByLabelText(ariaLabel)

        expect(labelPostit).not.toBeInTheDocument()
        expect(inputPostit).toBeInTheDocument()
      })
      it('Should type text solution in postit', async () => {
        const { setSolution } = events
        const { ariaLabel } = defaultProps

        const inputPostit = screen.getByLabelText(ariaLabel)
        await userEvent.type(inputPostit, 'abc')

        expect(setSolution).toBeCalledTimes(3)
        expect(inputPostit).toHaveValue('abc')
      })
    })
  })
  describe('Postit test with real context valitading', () => {
    const setupComponent = (props = defaultProps) =>
      render(
        <ExerciseStore>
          <Postit {...props} />
        </ExerciseStore>
      )

    describe('Postit readonly component', () => {
      it('Should render Postit component of readonly type', () => {
        const { asFragment } = setupComponent()

        expect(asFragment()).toMatchSnapshot()
      })
      it('Should make readonly the postit component', () => {
        const { text } = defaultProps
        setupComponent()

        const postit = screen.getByText(text)

        expect(postit).toHaveClass('postit--readonly')
      })
    })
    describe('Postit not readonly component', () => {
      beforeEach(() => {
        setupComponent({
          ...defaultProps,
          text: undefined,
          readonly: false,
        })
      })

      it('Should render Postit component of not readonly type', () => {
        const { asFragment } = setupComponent({
          ...defaultProps,
          text: undefined,
          readonly: false,
        })

        expect(asFragment()).toMatchSnapshot()
      })

      it('Should render label and input from postit', () => {
        const { htmlFor, ariaLabel } = defaultProps

        const labelPostit = screen.queryByLabelText(htmlFor)
        const inputPostit = screen.getByLabelText(ariaLabel)

        expect(labelPostit).not.toBeInTheDocument()
        expect(inputPostit).toBeInTheDocument()
      })
      it('Should type text solution in postit', async () => {
        const { ariaLabel } = defaultProps

        const inputPostit = screen.getByLabelText(ariaLabel)
        await userEvent.type(inputPostit, 'abc')

        expect(inputPostit).toHaveValue('abc')
      })
    })
  })
})
