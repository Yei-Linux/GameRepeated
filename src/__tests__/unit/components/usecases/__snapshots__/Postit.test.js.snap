// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Postit component Postit test with mocked context Postit not readonly component Should render Postit component of not readonly type 1`] = `
<DocumentFragment>
  <div
    class="postit--notreadonly container--pd0"
    data-testid="postitNotReadonly"
  >
    <input
      aria-label="postItElement"
      autocomplete="off"
      class="postit--dimensions postit__inputnumber--dimensions input input--primary input--border1 input--pd1"
      name="0"
      type="text"
      value="abc"
    />
  </div>
</DocumentFragment>
`;

exports[`The Postit component Postit test with mocked context Postit readonly component Should render Postit component of readonly type 1`] = `
<DocumentFragment>
  <div
    class="postit postit--dimensions postit--primary postit--readonly container--pd1"
    data-testid="postitReadonly"
  >
    PostIt Component
  </div>
</DocumentFragment>
`;

exports[`The Postit component Postit test with real context valitading Postit not readonly component Should render Postit component of not readonly type 1`] = `
<DocumentFragment>
  <div
    class="postit--notreadonly container--pd0"
    data-testid="postitNotReadonly"
  >
    <input
      aria-label="postItElement"
      autocomplete="off"
      class="postit--dimensions postit__inputnumber--dimensions input input--primary input--border1 input--pd1"
      name="0"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`The Postit component Postit test with real context valitading Postit readonly component Should render Postit component of readonly type 1`] = `
<DocumentFragment>
  <div
    class="postit postit--dimensions postit--primary postit--readonly container--pd1"
    data-testid="postitReadonly"
  >
    PostIt Component
  </div>
</DocumentFragment>
`;
